src/main/resources/hazelcast-config-schema.json

```diff
@@ -1373,20 +1373,535 @@
+            }
+          }
+        },
+        "failure-detector": {
+          "type": "object",
+          "description": "A failure detector is responsible to determine if a member in the cluster is unreachable or crashed. Please refer to https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#failure-detector-configuration for the failure detectors implemented by Hazelcast.",
+          "additionalProperties": false,
+          "properties": {
+            "icmp": {
+              "type": "object",
+              "description": "ICMP can be used in addition to the other detectors. It operates at layer 3 and detects network and hardware issues more quickly.",
+              "additionalProperties": false,
+              "properties": {
+                "enabled": {
+                  "type": "boolean",
+                  "default": false
+                },
+                "timeout-milliseconds": {
+                  "type": "integer",
+                  "minimum": 1,
+                  "default": 1000,
+                  "description": "Timeout in milliseconds before declaring a failed ping."
+                },
+                "fail-fast-on-startup": {
+                  "type": "boolean",
+                  "default": true,
+                  "description": "Cluster Member will fail to start if it is unable to action an ICMP ping command when ICMP is enabled. Failure is usually due to OS level restrictions."
+                },
+                "interval-milliseconds": {
+                  "type": "integer",
+                  "minimum": 1000,
+                  "default": 1000,
+                  "description": "Time in milliseconds between each ICMP ping."
+                },
+                "max-attempts": {
+                  "type": "integer",
+                  "minimum": 0,
+                  "default": 2,
+                  "description": "Maximum number of consecutive failed attempts before declaring a member suspect."
+                },
+                "parallel-mode": {
+                  "type": "boolean",
+                  "default": true,
+                  "description": "Run ICMP detection in parallel with the Heartbeat failure detector."
+                },
+                "ttl": {
+                  "type": "integer",
+                  "minimum": 0,
+                  "default": 255,
+                  "description": "Maximum number of times the IP Datagram (ping) can be forwarded, in most cases all Hazelcast cluster members would be within one network switch/router therefore default of 0 is usually sufficient."
+                }
+              }
+            }
+          }
+        },
+        "member-server-socket-endpoint-config": {
+          "type": "object",
+          "description": "Configures the server socket used for the member-member communication.",
+          "additionalProperties": false,
+          "$ref": "#/definitions/SocketEndpointConfig"
```

Comment:
Please put the "type" and "additionalProperties" to the SocketEndpointConfig definition. The problem with this current approach is that according to json schema, if an schema contains a "$ref", then every other property should be ignored. I'm happy to see that IntelliJ at least interprets the "description" properties adjacent to "$ref"s, but I don't expect much compatibility from validators.

Reply:
done